### DISCUSSION TITLE
Third party modules?

### DISCUSSION THREAD

**Discussion by hongkongkiwi** [finit-project/finit#270] [Q&A] [answered]
Whats the best way to submit a couple of extra plugins I wrote? Do you think it's worth keeping a separate repo for third party modules? (and bringing it in as a subrepo?).

Specifically I've created a mdevd & qmi-proxy plugin.

**Comment by hongkongkiwi** [comment]
The mdevd plugin is a bit tricky, I wanted to make sure that it runs first, like dbus.

But I had to modify the modprobe & hotplug plugin because I want to add the mdevd plugin as a dependency. Perhaps there's a way to set a "soft" dependency here so if it exists it runs first, otherwise it falls back to bootmisc dependency.

```
static plugin_t plugin = {
	.name = __FILE__,
	.hook[HOOK_BASEFS_UP] = {
		.cb  = setup
	},
#ifdef MDEVD_PLUGIN_COLDPLUG
	.depends = { "mdevd" },
#else
        .depends = { "bootmisc" },
#endif
};
```

**Comment by troglobit** [comment]
Nah, let's bring 'em in to finit main :)

We can just make sure to have them disabled by default. I've already scheduled mdevd #195 for the next release, so that one's a no-brainer.  Didn't know about QMI until today, but since it's an open source thing I wouldn't mind have that one also in Finit main.  I've considered a 3rd party plugin repo before, but submodules is a mess to manage release-wise, so we'll take that bridge when we get to it.

Regarding the mdevd dep. in hotplug and modprobe, you can just add it to the list, no need to #ifdef, if iẗ́s not available (not built) Finit will just skip the dependency.  (Very much how I got the idea for condition soft deps we discussed the other day :)

So, like this:

    .depends = { "bootmisc", "mdevd" },

**Comment by hongkongkiwi** [comment]
Oh excellent that should make it much easier! Here's the modules:
- [mdevd.c](https://github.com/hongkongkiwi/finit/blob/master/plugins/mdevd.c)
- [qmi-proxy.c](https://github.com/hongkongkiwi/finit/blob/master/plugins/qmi-proxy.c)

FYR qmi-proxy plugin is for libqmi, or specifically the qmicli cli tool for qmi_wwan.ko module so you can run multiple commands using the same lte device at once without worrying about allocating/releasing ids.

For the [hotplug.c](https://github.com/hongkongkiwi/finit/blob/master/plugins/hotplug.c) plugin, I have to used this to decide what to run:
```
#ifdef MDEVD_PLUGIN_COLDPLUG
#define MDEV_COLDPLUG_ARGS   ""
#else
#define MDEV_COLDPLUG_ARGS    "-s"
#endif
```
```
#ifdef MDEVD_PLUGIN_COLDPLUG
    path = which("mdevd-coldplug");
#else
    path = which("mdev");
#endif
    if (path) {
      /* Embedded Linux systems usually have BusyBox mdev */
      if (debug)
        touch("/dev/mdev.log");

      snprintf(cmd, sizeof(cmd), "%s %s", path, MDEV_COLDPLUG_ARGS);
      free(path);

      run_interactive(cmd, "Populating device tree");
    }
```

but I was thinking, we probably shouldn't even need the compile time define's if we are able to detect if mdevd plugin has been loaded or not, this would account for the dynamic plugins which are not there at compile time. Is there a way to check a module is loaded at run time? 

If the mdevd plugin has been loaded then use mdevd-coldplug, otherwise use normal mdev

**Reply by troglobit** [reply]
Cool thanks, I'll look into it when I get some time over!  For now I'll just add a separate issue to the tracker for the qmi-proxy plugin.

Hmm, maybe we could refactor `sig.c:do_iterate_proc()` into a `proc_find()`, then we could just check if mdevd is running, and have a different code path.  That's better than compile time, as you say.

**Comment by troglobit** [comment, marked as answer]
Concluding this with the resulting repo -> https://github.com/troglobit/finit-plugins :-)

**Reply by troglobit** [reply]
mdevd plugin now merged to Finit master, and the qmi-proxy plugin I put in the finit-plugins repo instead: https://github.com/troglobit/finit-plugins/commit/f18a5802adbf0907c95742a272d05612dce239c9

For an example of how to integrate in Buildroot, see https://github.com/troglobit/myLinux/tree/main/package/finit-plugins
